components/brave_ads/browser/ads_service_impl.cc

```diff
@@ -249,13 +252,19 @@ AdsServiceImpl::AdsServiceImpl(
 
   GetDeviceIdAndMaybeStartBatAdsService();
 
-  g_brave_browser_process->resource_component()->AddObserver(this);
+  if (g_brave_browser_process->resource_component()) {
+    g_brave_browser_process->resource_component()->AddObserver(this);
+  } else {
+    CHECK_IS_TEST();
+  }
 
   rewards_service_->AddObserver(this);
 }
 
 AdsServiceImpl::~AdsServiceImpl() {
-  g_brave_browser_process->resource_component()->RemoveObserver(this);
+  if (g_brave_browser_process->resource_component()) {
```

Comment:
this is a layering violation anyway so can we change this to pass in `g_brave_browser_process->resource_component()` to the constructor?

Reply:
cc @tmancey https://github.com/brave/brave-core/pull/18557/files#r1681672486